src/coreclr/jit/gentree.cpp

```diff
@@ -1060,6 +1060,29 @@ bool GenTreeCall::HasSideEffects(Compiler* compiler, bool ignoreExceptions, bool
         return true;
     }
 
+    // Consider array allocators side-effect free for constant length (if it's not negative and fits into i32)
+    if (helperProperties.IsAllocator(helper))
+    {
+        GenTree* arrLen = compiler->getArrayLengthFromAllocation((GenTree*)this DEBUGARG(nullptr));
+        // if arrLen is nullptr it means it wasn't an array allocator
+        if ((arrLen != nullptr))
+        {
+            arrLen = arrLen->OperIsPutArg() ? arrLen->gtGetOp1() : arrLen;
+            if (arrLen->IsIntCnsFitsInI32())
+            {
+                // Array.MaxLength
+                const ssize_t MaxArraySize = 0x7FFFFFC7;
+
+                ssize_t cns = arrLen->AsIntConCommon()->IconValue();
+                if ((cns >= 0) && (cns <= MaxArraySize))
+                {
+                    // Actually the max allowed length is smaller than
```

Comment:
than?

Reply:
Removed 🙂 it was before I introduced `MaxArraySize = 0x7FFFFFC7;` constant. Some tests failed because they did `new int[int.MaxValue]` and expected OutOfMemoryException 😄